src/vsr/replica.zig

```diff
@@ -4479,112 +4360,195 @@ pub fn Replica(
+            // This guarantees that if the old leader possibly committed the operation, then the
+            // new leader will also commit the operation.
+            if (commit_max < self.commit_max and self.commit_min == self.commit_max) {
                 log.debug("{}: {s}: k={} < commit_max={} and commit_min == commit_max", .{
                     self.replica,
                     method,
-                    k,
+                    commit_max,
                     self.commit_max,
                 });
             }
-            assert(k >= latest.commit);
-            assert(k >= self.commit_max - std.math.min(config.pipeline_max, self.commit_max));
+            assert(commit_max >=
+                self.commit_max - std.math.min(config.pipeline_max, self.commit_max));
 
             assert(self.commit_min <= self.commit_max);
             assert(self.op >= self.commit_max or self.op < self.commit_max);
 
-            self.op = latest.op;
+            const previous_op = self.op;
+            const previous_commit_max = self.commit_max;
+
+            self.op = op;
+
             // Crucially, we must never rewind `commit_max` (and then `commit_min`) because
             // `commit_min` represents what we have already applied to our state machine:
-            self.commit_max = std.math.max(self.commit_max, k);
+            self.commit_max = std.math.max(self.commit_max, commit_max);
 
             assert(self.commit_min <= self.commit_max);
-            assert(self.op >= self.commit_max);
+            assert(self.commit_max <= self.op);
 
-            // Do not set the latest op as dirty if we already have it exactly:
-            // Otherwise, this would trigger a repair and delay the view change, or worse, it would
-            // prevent us from assisting another replica to recover when we do in fact have the op.
-            if (self.journal.has(latest)) {
-                log.debug("{}: {s}: latest op exists exactly", .{ self.replica, method });
-            } else {
-                self.journal.set_header_as_dirty(latest);
+            self.journal.remove_entries_from(self.op + 1);
+
+            log.debug("{}: {s}: view={} op={}..{} commit={}..{}", .{
+                self.replica,
+                method,
+                self.view,
+                previous_op,
+                self.op,
+                previous_commit_max,
+                self.commit_max,
+            });
+        }
+
+        fn set_log_from_do_view_change_messages(self: *Self) void {
+            const view_normal_canonical = self.set_op_and_commit_max_from_do_view_change_messages();
+
+            // First, set all the canonical headers from the replica(s) with highest `view_normal`:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    var view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
```

Comment:
The view during which the replica's status was last normal, has to be before the view of the current view change, since we are advancing to a newer view. Is that a clear explanation?

I've added a comment to this effect.

Reply:
Gotcha, this message is always generated during a view change, which means it's view_normal is in the past. I was mixed up about the context.